core/src/impl/Kokkos_Core.cpp

```diff
@@ -166,36 +166,7 @@ int get_ctest_gpu(const char* local_rank_str) {
   return std::atoi(id.c_str());
 }
 
-namespace {
-
-bool is_unsigned_int(const char* str) {
-  const size_t len = strlen(str);
-  for (size_t i = 0; i < len; ++i) {
-    if (!isdigit(str[i])) {
-      return false;
-    }
-  }
-  return true;
-}
-
-void initialize_backends(const InitArguments& args) {
-// This is an experimental setting
-// For KNL in Flat mode this variable should be set, so that
-// memkind allocates high bandwidth memory correctly.
-#ifdef KOKKOS_ENABLE_HBWSPACE
-  setenv("MEMKIND_HBW_NODES", "1", 0);
-#endif
-
-  // Protect declarations, to prevent "unused variable" warnings.
-#if defined(KOKKOS_ENABLE_OPENMP) || defined(KOKKOS_ENABLE_THREADS) || \
-    defined(KOKKOS_ENABLE_OPENMPTARGET) || defined(KOKKOS_ENABLE_HPX)
-  const int num_threads = args.num_threads;
-#endif
-#if defined(KOKKOS_ENABLE_THREADS) || defined(KOKKOS_ENABLE_OPENMPTARGET)
-  const int use_numa = args.num_numa;
-#endif
-#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_ROCM) || \
-    defined(KOKKOS_ENABLE_HIP)
+int get_gpu(const InitArguments& args) {
```

Comment:
Why did you define it outside the anonymous namespace?

Reply:
I just followed the example of get_ctest_gpu() that was also in that file:

* https://github.com/kokkos/kokkos/blob/0e51c8de02e5d9dd7bc87b99f2bb42ba7b7eae1d/core/src/impl/Kokkos_Core.cpp#L80

Also, implementation functions like this should be independently unit tested.  Having this in Kokkos::Impl instead of an anonymous namespace would allow that.

Reply:
Right that's why `get_ctest_gpu` is outside and we test it in `core/unit_test/TestCTestDevice.cpp` using a forward declaration.

But `get_gpu` does not have dedicated tests and can (and IMO should) be within that anonymous namespace.

Reply:
It is find to move git_gpu() into an anonymous namespace for now.  When someone gets around to unit testing it, it can be moved back to the Kokkos::Impl namespace.

Reply:
>But get_gpu does not have dedicated tests and can (and IMO should) be within that anonymous namespace.

@dalg24, completed in https://github.com/kokkos/kokkos/pull/3048/commits/7c0f96729830a9b580a45032376fc7493d741482.

Reply:
I did not test this last change.  I guess we will test the Kokkos PR testing :-)